src/github.com/getlantern/peerscanner/host.go

```diff
@@ -80,33 +87,75 @@ func (h *host) String() string {
  * API for interacting with host
  ******************************************************************************/
 
-// newHost creates a new host for the given name, ip and optional DNS record.
-func newHost(name string, ip string, record *cloudflare.Record) *host {
-	// Cache TLS sessions
-	clientSessionCache := tls.NewLRUClientSessionCache(1000)
-
+// newHost creates a new host for the given name, ip and optional DNS records.
+func newHost(name string, ip string, port string, cflRecord *cloudflare.Record, dspRecord *dnsimple.Record) *host {
 	h := &host{
 		name:         name,
 		ip:           ip,
-		record:       record,
+		port:         port,
+		cflRecord:    cflRecord,
+		dspRecord:    dspRecord,
 		resetCh:      make(chan string, 1000),
 		unregisterCh: make(chan interface{}, 1),
 		statusCh:     make(chan chan *status, 1000),
+		initCfrCh:    make(chan interface{}, 1),
+	}
+
+	if h.isFallback() {
+
+		h.cflGroups = map[string]*cflGroup{
+			RoundRobin: &cflGroup{subdomain: RoundRobin},
+			Fallbacks:  &cflGroup{subdomain: Fallbacks},
+			Peers:      &cflGroup{subdomain: Peers},
+		}
+		h.dspGroups = map[string]*dspGroup{
+			RoundRobin: &dspGroup{subdomain: RoundRobin},
+			Fallbacks:  &dspGroup{subdomain: Fallbacks},
+			Peers:      &dspGroup{subdomain: Peers},
+		}
+		country := fallbackCountry(name)
+		if country != "" {
+			// Add host to country-specific rotation
+			h.cflGroups[country] = &cflGroup{subdomain: country}
+			h.dspGroups[country] = &dspGroup{subdomain: country}
+		}
+	} else {
+		log.Errorf("Somehow adding peer host? %v (%v)", name, ip)
+	}
+
+	return h
+}
+
+func (h *host) resetProxiedClient(port string) {
+
+	var dial func(addr string) (net.Conn, error)
+	if port == "80" {
+		dialer := net.Dialer{Timeout: dialTimeout}
```

Comment:
> TBH I don't see any correctness hazard or clarity problem here. On that respect, dialer might as well be global. It'd be no worse than (cfl|cfr|dsp)util, and those are used less often.

I just think state should be avoided wherever possible, and it doesn't seem like it buys us anything here -- I'm certainly not condoning the other globals by questioning this one!

> We are not storing any state (as in, anything that changes) in proxiedClient. 

Well, proxiedClient itself is updated with it's new state -- each proxied client itself is immutable I suppose sure.

> resetProxiedClient does what (IIUC) you want newClient to do, except, as a slight convenience, it also sets the client to the host when it could be created successfully. 

I would actually do the following:

1. Remove proxiedClient from the host struct
1. Kill the doIsAbleToProxy method entirely
1. Create a new method for creating the client that just gets called when we need the client, which is currently further down in reallyDoIsAbleToProxy

That also allows us to only deal with the client if and when we actually need it -- i.e. once the raw connection test has happened. It's not a huge deal, but otherwise we're doing all this to me confusing gymnastics just to reset the proxiedClient to have the correct dialer, which is hardly an optimization to begin with.

```go
func (h *host) newProxiedClient(port string) http.Client {
	var dial func(addr string) (net.Conn, error)
	if port == "80" {
		dialer := net.Dialer{Timeout: dialTimeout}
		dial = func(addr string) (net.Conn, error) {
			return dialer.Dial("tcp", h.ip+":80")
		}
	} else if port == "443" {
		dial = func(addr string) (net.Conn, error) {
			return tlsdialer.DialWithDialer(&net.Dialer{
				Timeout: dialTimeout,
			}, "tcp", h.ip+":443", true, &tls.Config{
				InsecureSkipVerify: true,
				// Cache TLS sessions
				ClientSessionCache: tls.NewLRUClientSessionCache(1000),
			})
		}
	} else {
		log.Errorf("Unsupported port: %v", port)
		return
	}

	return &http.Client{
		Transport: &http.Transport{
			Dial: func(network, addr string) (net.Conn, error) {
				return enproxy.Dial(addr, &enproxy.Config{
					DialProxy: dial,
					NewRequest: func(upstreamHost string, method string, body io.Reader) (req *http.Request, err error) {
						return http.NewRequest(method, "http://"+h.ip+"/", body)
					},
					OnFirstResponse: func(resp *http.Response) {
						h.reportedHostMutex.Lock()
						h.reportedHost = resp.Header.Get(enproxy.X_ENPROXY_PROXY_HOST)
						h.reportedHostMutex.Unlock()
					},
				})
			},
			DisableKeepAlives: true,
		},
		Timeout: requestTimeout,
	}
}
```